### ISSUE TITLE
Bedgraph/Bigwig Input Files Not Normalized Properly

### ISSUE THREAD

**Issue by mgfield** [simonvh/fluff#64]
Inputting a different bedgraph file for two different conditions results in a normalization in which the heatmap and bandplot show similar "max" intensity.  I believe this is due to the normalization based on read counts, as this issue does not occur when I use the -r option with bam files.  It also does not occur when using deeptools with default settings from the bigwigs, so it has to be a normalization issue.

**Comment by simonvh** [comment]
I understand your problem. With bedgraph files, fluff uses normalized (or essentially scaled) data only for clustering. This is not based on read counts, but based on a specific percentile of the values per bin. However, the tracks are shown in the heatmap using the original data. In the heatmap the color scale is adapted per track and set to the 90% percentile. You can change this behaviour for `fluff heatmap` by specifying a value for the `-s` parameter. This is the maximum value for the color scale, which will be used for each track. In this way all the tracks will use the same maximum value. I think this should fix the issue.

For `fluff bandplot` the solution is something similar. The default behaviour is once again to set the scale for each track separately. The `-s` works differently here. You can specify which tracks should use the same scale. For instance, `-s 1:3,4:6` would group the first three tracks and the second three tracks.

I am sorry that these options are not that well documented. Hopefully this should solve your problem, please let us know.

**Comment by mgfield** [comment]
Thanks Simon, but I have played around with these options.  Unfortunately,
they don't solve the issue.  Let me show you in more detail with an example.

I'll use bandplot as the example.  I took 2 bigwig files and 2 bam files
from the same dataset.  The bigwig files have already been normalized in a
special manner using a different program, so that is why I prefer to use
them than the bams.  For both these plots, I set -s 1:2.  The only
difference is that the bam plot used -r, and the bedgraph plot didn't
because you can't use RPKM normalization for bedgraphs.

In the bedgraph plot (top figure), it looks like both bedgraphs have almost
exactly the same clusters, where the top of the TSS is normalized to be the
exact same value in the two cases.  The bottom plot (bam files) shows a
clear increase in the TSS region in cluster 2 and 3, which is what I expect
to see and the actual data shows. Thus, I think the normalization to read
counts for the bandplot and heatmap is the issue.  Perhaps, turning this
feature off and not using any normalization for bedgraph and bigwig files
that are already normalized is the solution.  Can this be implemented
easily?

Bedgraph plot:

[image: Inline image 1]

Bam plot:
[image: Inline image 2]

Thanks,
Matt

**Comment by simonvh** [comment]
Thanks for the example. To be clear, this is not a feature. There should be no normalization for bedGraph or bigWig files as these are usually already normalized. So, this is either a bug or it has a different reason somehow and it is not expected behaviour. We'll look into it.

**Comment by mgfield** [comment]
Great.  I really like the ease of use and graphics of fluff, and I would
like to use this in our publication.  Do you think you can fix this
quickly?  Can you give me an idea of your timeline because we will have to
move to a different tool if it will take awhile, as we are putting together
the figures now for a couple of different publications.  Thanks.

**Comment by simonvh** [comment]
If we can track down the issue it should be fixed soon. Can you give me the exact commands you use? I don't seem to have any trouble with bigWigs. For example, I created two files with normally distributed values (mean 1 and mean 10):

```
$ head test_1.wig
chr1	1	2	-0.0490561280896
chr1	2	3	1.15815837766
chr1	3	4	1.41438525232
chr1	4	5	2.88554513664
chr1	5	6	1.4490469068
chr1	6	7	1.84966401605
chr1	7	8	1.0143481173
chr1	8	9	0.967501668971
chr1	9	10	0.0704036313677
chr1	10	11	1.47888771668
```

```
$ head test_10.wig
chr1	1	2	9.35161381369
chr1	2	3	10.9477579801
chr1	3	4	8.98617755124
chr1	4	5	9.58836292139
chr1	5	6	10.1108579171
chr1	6	7	9.16536474231
chr1	7	8	8.60085405484
chr1	8	9	11.4078945161
chr1	9	10	9.68043969372
chr1	10	11	9.35035608756
```

If I convert these files to bigWig, using UCSC wigToBigWig, and use them to create a bandplot, I get the expected output.

```
$ fluff bandplot -f test.bed -d test_1.bw test_10.bw -o bandplot.png -b 10 -s 1:2
```
![bandplot](https://cloud.githubusercontent.com/assets/281252/25650599/58f096b0-2fde-11e7-9916-e64b5b519b8c.png)

**Comment by mgfield** [comment]
When I use my bigwig files with your program, I get the error shown below.
I think it might be because the chromosome sizes file has sizes that are
less by one or two nucleotides than the actual reads for certain
chromosomes, as I saw that warning message when using other programs.  But,
I am not exactly sure, and I don't get any error using bigwigs with
deeptools.  However, when I convert the bigwigs to bedgraph and use fluff,
I for some reason no longer get the error.  But, I then have the
normalization problem.  My command with bedgraph files is below.  So, based
on this..

1. Can you test your script with bedgraph files to see if that is the issue?

2. Can you help figure out why the bigwig files aren't working?

fluff bandplot -f tssDown_clusters.bed \
-d AC_KO_average.bdg AC_WT_average.bdg \
-s 1:2 -S \
-o bandplot1_2TSSDownS_bdg


Bigwig Error:

Traceback (most recent call last):
  File "/projects/med/hlab/software/virtualpython/ENV/bin/fluff", line 327,
in <module>
    args.func(args)
  File
"/projects/med/hlab/software/virtualpython/ENV/lib/python2.7/site-packages/fluff/commands/bandplot.py",
line 51, in bandplot
    data = load_cluster_data(clust_file, datafiles, bins, rpkm, rmdup,
rmrepeats, fragmentsize=fragmentsize)
  File
"/projects/med/hlab/software/virtualpython/ENV/lib/python2.7/site-packages/fluff/fluffio.py",
line 279, in load_cluster_data
    fragmentsize=fragmentsize)
  File
"/projects/med/hlab/software/virtualpython/ENV/lib/python2.7/site-packages/fluff/fluffio.py",
line 423, in get_binned_stats
    readlength = track.read_length()
  File
"/projects/med/hlab/software/virtualpython/ENV/lib/python2.7/site-packages/fluff/fluffio.py",
line 138, in read_length
    for read in self.track:
  File "pybedtools/cbedtools.pyx", line 772, in
pybedtools.cbedtools.IntervalIterator.__next__
(pybedtools/cbedtools.cxx:11001)
  File "pybedtools/cbedtools.pyx", line 637, in
pybedtools.cbedtools.create_interval_from_list
(pybedtools/cbedtools.cxx:9068)
IndexError: list index out of range

Thanks,
Matt

**Comment by mgfield** [comment]
I just figured out the problem with the bedgraphs.  If the file is named
.bedGraph the error occurs that I was talking about, the normalization is
incorrect.  If the file is named .bdg, it works fine.

I thought this when I edited my command above and realized that my bedgraph
files were not named .bdg but were named .bedGraph.  When I changed
.bedGraph to .bdg, it worked fine.

I still am not sure why the bigwigs don't work though.

Thanks,
Matt
